src/Native/LDANative/lda_engine.cpp

```diff
@@ -0,0 +1,1103 @@
+                token_sweeped += sampler_.InferOneDoc(doc.get());
+            }
+            atomic_stats_->num_tokens_clock_ += token_sweeped;
+
+            process_barrier_->wait();
+            double iter_end = lda::get_time();
+
+            if (thread_id == 0)
+            {
+                double seconds_this_iter = iter_end - iter_start;
+
+                printf("Iter: %04d", iter);
+                std::cout
+                    << "\tThread = " << thread_id
+                    << "\tTokens: " << atomic_stats_->num_tokens_clock_
+                    << "\tTook: " << seconds_this_iter << " sec"
+                    << "\tThroughput: "
+                    << static_cast<double>(atomic_stats_->num_tokens_clock_) / (seconds_this_iter) << " token/(thread*sec)"
+                    << std::endl;
+
+            }
+
+            process_barrier_->wait();
+
+            if (compute_ll_interval_ != -1 && (iter % compute_ll_interval_ == 0 || iter == burnin_iterations_ - 1))
+            {
+                EvalLogLikelihood(false, thread_id, iter, sampler_);
+            }
+        }
+
+        double total_end = lda::get_time();
+        printf("thread_id = %d, Total time for burnin iterations : %f sec.\n", thread_id, total_end - total_start);
+    }
+
+    void LdaEngine::AllocateDataMemory(int num_document, int64_t corpus_size)
+    {
+        data_block_->Allocate(num_document, corpus_size);
+    }
+
+    void LdaEngine::AllocateModelMemory(const LDADataBlock* data_block)
+    {
+        model_block_->InitFromDataBlock(data_block, V_, K_);
+
+        global_word_topic_table_.resize(V_);
+
+        for (int i = 0; i < V_; ++i)
+        {
+            global_word_topic_table_[i] = model_block_->get_row(i, nullptr);
+        }
+    }
+
+    void LdaEngine::AllocateModelMemory(int num_vocabs, int num_topics, int64_t nonzero_num)
+    {
+        model_block_->Init(num_vocabs, num_topics, nonzero_num);
+
+        global_word_topic_table_.resize(num_vocabs);
+
+        for (int i = 0; i < num_vocabs; ++i)
+        {
+            global_word_topic_table_[i] = model_block_->get_row(i, nullptr);
+        }
+    }
+
+    void LdaEngine::AllocateModelMemory(int num_vocabs, int num_topics, int64_t mem_block_size, int64_t alias_mem_block_size)
+    {
+        model_block_->Init(num_vocabs, num_topics, mem_block_size, alias_mem_block_size); //memory allocated here
+
+        global_word_topic_table_.resize(num_vocabs);
+        global_summary_row_.resize(K_, 0);
+
+        //each value inside the global_word_topic_table_ will be set while call SetWordTopic()
+    }
+
+    int LdaEngine::FeedInData(int* term_id, int* term_freq, int32_t term_num, int32_t vocab_size)
+    {
+        // bool b_ret = true;
+        if (V_ == 0) //number vocab could be set in allocating model memory function
+            V_ = vocab_size;
+
+        //data_block represent for one doc
+        return data_block_->Add(term_id, term_freq, term_num);
+        // return b_ret;
```

Comment:
>// return b_ret; [](start = 8, length = 16)

can we clean it? and commented definition above. Same for FeedInDataDense function